README.md
# raspi-headless-setup

Configure a Raspberry Pi OS microSD for running in headless mode.

## Synopsis

```sh
raspi-headless-setup.sh [options] SSID
```

## Description

This program is used to configure a Raspberry Pi to run headless.  A
headless Raspberry Pi is one that does not need a monitor, keyboard or
mouse. Interaction with it can be done over the network.  For example,
using the command line over SSH, or via a graphical session over
VNC. The only wired connection to the Raspberry Pi needed is the
power. Either Wi-Fi and/or a wired connection is required.


This program is run on the computer used to prepare a microSD card for
the Raspberry Pi. After imaging the Raspberry Pi OS onto the microSD
card, remount the microSD card and run this program.

This program can configure:

- Wi-Fi settings;
- SSH server;
- default "pi" user account password;
- default "pi" user account adding a SSH public key to .ssh/authorized_keys;
- hostname;
- timezone;
- locale;
- VNC server; and
- force the Raspberry Pi to  assume there is a HDMI display with a
  particular resolution.

### Options

#### Default user account credentials

- `-p | --pi-password` passwd

    Password for the "pi" user account. This can be used to change the
    default password of "raspberry".

    For better security, provide the password in a file with the `-P`
    option.

- `-P | --pi-password-file` filename

    Read the "pi" user account password from a file. This can be used
    to change the default password of "raspberry".

- `-k | --pi-ssh-pubkey` pubkey_file

    Create or add the contents of the _pubkey_file_ to the
    _.ssh/authorized_keys_ file for the "pi" user account.  By
    default, it will use the _~/.ssh/id_rsa.pub_ file of the account
    that invoked it, if it exists and `--nopubkey` has not been
    specified.

- `-K | --pubkey-only`

    Prohibit the use of passwords for SSH authentication.  By default,
    either passwords or (if a public key is provided) SSH public keys
    can be used for authentication.  This option cannot be used if
    there is no public key configured for the "pi" user account
    (i.e. it cannot be used if `--no-pubkey` is used, or no public key
    has been provided via the `--pi-ssh-pubkey` option or its default
    value). Use of this option improves security.

- `--no-pubkey`

    Do not configure the _.ssh/authorized_keys_ file for the "pi" user
    account.  If there is a default public key file, this option will
    ensure it will not be used, because without it the authorized keys
    file will be configured using that default file, even though the
    `--pi-ssh-pubkey` was not specified.

#### Other defaults

- `-n | --hostname` name

    Set the hostname of the Raspberry Pi. This can be used to change
    the hostname from the default value of "raspberrypi".

- `-t | --timezone` tz_name

    Set the timezone. Timezone names are in the form of
    "Area/Location" and are case sensitive.

    The timezone names should be from the [tz
    database](https://en.wikipedia.org/wiki/List_of_tz_database_time_zones).
    For a list of available timezones, on the Raspberry Pi run:
    `timedatectl list-timezones`

    This program tries to determine a default value from the local
    computer. If it is not set, the Raspberry Pi's default value is
    British Standard Time.

- `-l | --locale` locale

    Set the locale.

    For available locale names, look inside _/etc/locale.gen_ on the
    Raspberry Pi.

    This program tries to determine a default value from the local
    computer. If it is not set, the Raspberry Pi's default value is
    "en_GB.UTF-8".

- `--no-locale`

    Do not attempt to set the locale.

    Leave the locale unchanged (en_GB is the default on the Raspberry
    Pi).  This option can be useful for quick testing, because setting
    the locale takes the most time when the Raspberry Pi boots for the
    first time (about an extra 1.5 minutes).

#### Wi-Fi options

- `-c | --country` code

    Two letter ISO 3166-1 country code for setting up Wi-Fi
    frequencies.  This is mandatory if Wi-Fi is being configured.  The
    value is case-insensitive.

    This program tries to determine a default value from the locak
    computer.

- `-s | --ssid-passphrase` plaintext_passphrase

    The plaintext version of the Wi-Fi passphrase.  A WPA-PSK Wi-Fi
    passphrase must be at least 8 characters long.

    For better security, provide the passphrase in a file with the
    `-S` option.

- `-S | --ssid-passphrase-file` file

    Read the plaintext version of the Wi-Fi passphrase from a file.

- `-k | --psk` hex_PSK_value

    Wi-Fi passphrase and SSID hashed into the PSK, and represented in
    hexadecimal. This is the preferred way of providing the Wi-Fi
    passphrase, since it avoids storing the plaintext version in the
    configuration file. If this value is not known, provide
    the passphrase in plaintext.

- `--no-wifi`

    Do not configure the Wi-Fi. Use this option if Wi-Fi is not being
    used, and only a wired network connection will be used.

#### VNC options

- `-g | --vnc-password` vnc_password

    provided the VNC password on the command line. It must be between
    6 to 8 characters long.

    This program does not prompt for the VNC password, but will use
    the default value of "password" if a VNC password is not provided.

- `-G | --vnc-password-file` file

    Read the VNC password from a file.

- `--no-vnc`

    Do not attempt to configure the VNC server.

- `-r | --hdmi` resolution

    The hdmi_group and hdmi_mode that defines the resolution of the
    HDMI display.

    The resolution must be either "0" to autodetect the resolution
    (only useful if a physical monitor is connected to the HDMI port)
    or the _hdmi_group_ and _hdmi_mode_ separated by a slash.
    For example `2/35` is a resolution of 1280x1024 pixels.
    There is a list of [HDMI groups and
modes](https://www.raspberrypi.org/documentation/configuration/config-txt/video.md)
    that are supported by the Raspberry Pi.

- `--no-hdmi`

    Do not configure HDMI.

#### Other options

- `-b | --boot` mount_point

    The directory where the Raspberry Pi boot partition has been
    mounted.

    This program tries to automatically determine the boot partition,
    by looking for well-known mount points (e.g. _/Volumes/boot_ on
    macOS). Use this option if it cannot detect it or detects the
    wrong one.

#### General options

- `-q | --quiet`

    No output unless there is an error.

- `-v | --verbose`

    Output extra information when running.

- `--version`

    Output the version information and exit.

- `-h | --help`

    Output a short help message and exit.

### Arguments

- SSID

    The SSID for the Wi-Fi network. This value is case sensitive.

    Note: the configuration works with either non-hidden or hidden SSIDs.

## Examples

### Step 1: image the operating system

Image a copy of the Raspberry Pi OS onto a microSD card. If the
graphical desktop interface is needed, use the full version. If only
the command line intereface is needed, the lite version can be used.

Eject and reinsert the microSD card. That should remount its _boot_
partition, which this program will add/modify files to.

### Step 2: run the setup program

To use a Wi-Fi network, the following will prompt for the Wi-Fi/SSID
password:

```sh
./raspi-headless-setup.sh "My-WiFi-SSID-name"
```

If using a wired network connection:

```sh
./raspi-headless-setup.sh --no-wifi
```

Other options can be specified. Normally, you would want to at least
change the hostname and the password for the default account

```sh
./raspi-headless-setup.sh --hostname mypi --pi-password-file password.txt "My-WiFi-SSID-name"
```

The Raspberry Pi will then have the network address of "mypi.local".

### Step 3: boot Raspberry Pi

Eject the microSD card.  Insert it into a Raspberry Pi and turn on the
power to it.

Wait for it to boot.  This first boot will take longer than usual.
How long it takes will depend on the Raspberry Pi hardware (e.g for a
Raspberry Pi Zero W, the first boot takes 4.5 minutes and subsequent
boots take about 1 minute).

### Step 4: SSH to the Raspberry Pi

Unless they were changed, the Raspberry Pi will have the hostname of
"raspberrypi" and the default "pi" user account has "raspberry" as the
password.

Connect to it using SSH:

```sh
ssh pi@raspberrypi.local
```

Note: the Raspberry Pi should be assigned a dynamic IP address by
DHCP, and it can be contacted using the fully qualified domain name of
"raspberrypi.local", because the Raspberry Pi supports mDNS.  If the
SSH client computer does not support mDNS, the IP address that was
assigned to the Raspberry Pi will have to be discovered and used
instead. Most modern operating systems have support for mDNS.

### Step 5: use the Raspberry Pi

If the VNC server is configured, create a SSH tunnel to the Raspberry
Pi and use a VNC client over the tunnel. For security, external
connections to the VNC server are rejected.

Further configurations can be done using the graphical _Raspberry Pi
Configuration_ application, or by running `sudo raspi-config`.


## Requirements

This program runs on Unix-like environments, such as macOS and Linux
(including the Raspberry Pi OS).

It has been tested on images of the Raspberry Pi OS, Buster
(2021-05-07); both the lite version and the version with desktop.

## Troubleshooting

### Cannot resolve raspberrypi.local

Possible causes:

- The Raspberry Pi is not connected to the network:

    - It does not have hardware support for Wi-Fi;

    - The Wi-Fi password is incorrect;

    - The configured SSID is only available on 5 GHz Wi-Fi and the
      Raspberry Pi only has hardware support for 2.4 GHz Wi-Fi; or

    - The network cable is not plugged in.

- The client computer does not support mDNS, and therefore could not
  resolve the _.local_ network name.

- The network's DHCP server did not assign the Raspberry Pi an IP
  address.

- The hostname was changed, and is no longer "raspberrypi".

## Known issues

This program cannot disable Wi-Fi or the SSH server, if they have
previously been configured. It can only be used to enable or change
the Wi-Fi configurations; and/or to enable the SSH server.

## Files

This program creates or modifies these files on the boot partition:
_wpa_supplicant.conf_, _ssh_, _config.txt_ and _cmdline.txt_.

It also creates its own _headless_setup.sh_ script, which is executed
by the _cmdline.txt_ on boot. After successfully running, it is
deleted. That script contains the passwords and Wi-Fi passphrase, so
it should be kept confidential.

## See also

- GitHub repository for [raspi-headless-setup](https://github.com/hoylen/raspi-headless-setup).


CHANGELOG.md
# raspi-headless-setup Changelog

## 1.2.0

- Added `--pubkey-only` option to disable SSH passwords for better security.
- Fixed missing Wi-Fi passphrase not prompted for bug.

## 1.1.0

- Added `--no-locale` to speed up initial boots when correct locale not reqired.
- Combined headless setup script and VNC setup script into one script.

## 1.0.0

- Initial release.


pi-headless-setup.sh
#!/bin/bash
#
# Configure a Raspberry Pi OS imaged microSD to have Wi-Fi and SSH running.
#
# This script is run on the computer used to setup the microSD card.
# Run it after imaging the Raspberry Pi OS onto it and before it is
# inserted into the Raspberry Pi.
#
# Copyright (C) 2021, Hoylen Sue.
#================================================================

PROGRAM='pi-init-setup'
VERSION='1.0.0'

EXE=$(basename "$0" .sh)
EXE_EXT=$(basename "$0")

#----------------------------------------------------------------
# Error handling

# Bash strict mode: http://redsymbol.net/articles/unofficial-bash-strict-mode/

# Exit immediately if a simple command exits with a non-zero status.
#   Trap ERR for better error messages than "set -e" gives (but ERR only
#   works for Bash and unlike "set -e" it doesn't propagate into functions.
#   Can't figure out which command failed? Run using "bash -x".
set -e
trap 'echo $EXE: aborted; exit 3' ERR

set -u # fail on attempts to expand undefined environment variables
set -o pipefail # prevents errors in a pipeline from being masked

#----------------------------------------------------------------
# Constants

#----------------
# Default HDMI display resolution

DEFAULT_RESOLUTION='2/35' # 1280x1024

#----------------
# Default country (empty string means it must be provided on the command line)
# Can be changed on the command line with the --country option.

# Try to detect from LANG environment variable
DEFAULT_COUNTRY=
if echo $LANG | grep -q '^[a-z][a-z]_[A-Z][A-Z]\.' ; then
  # en_AU.UTF-8 -> AU
  DEFAULT_COUNTRY=$(echo $LANG | sed 's/^[a-z][a-z]_\(..\)\..*/\1/')
fi

# If the above does not work, and you are tired of always specifying the
# country as a command line option, hard code the default value.

# DEFAULT_COUNTRY=AQ

#----------------------------------------------------------------
# Constants

# hdmi_mode limits from:
# https://www.raspberrypi.org/documentation/configuration/config-txt/video.md

HDMI_MODE_MAX1=107
HDMI_MODE_MAX2=86

#----------------------------------------------------------------
# Error handling

# Exit immediately if a simple command exits with a non-zero status.
# Better to abort than to continue running when something went wrong.
set -e

#----------------------------------------------------------------
# Command line arguments

WIFI_COUNTRY=$DEFAULT_COUNTRY
WIFI_SSID=
WIFI_PASSWORD_PLAINTEXT=
WIFI_PASSWORD_PSK=
NO_WIFI=
RESOLUTION=$DEFAULT_RESOLUTION
BOOT_DIR=
QUIET=
VERBOSE=
SHOW_VERSION=
SHOW_HELP=

while [ $# -gt 0 ]
do
  case "$1" in
    -c|--country)
      if [ $# -lt 2 ]; then
        echo "$EXE: usage error: -c or --country missing value" >&2
        exit 2
      fi
      WIFI_COUNTRY="$2"
      shift # past option
      shift # past argument
      ;;
    -p|--password)
      if [ $# -lt 2 ]; then
        echo "$EXE: usage error: -p or --password missing value" >&2
        exit 2
      fi
      WIFI_PASSWORD_PLAINTEXT="$2"
      shift # past option
      shift # past argument
      ;;
    -k|--psk)
      if [ $# -lt 2 ]; then
        echo "$EXE: usage error: -k or --psk missing value" >&2
        exit 2
      fi
      WIFI_PASSWORD_PSK="$2"
      shift # past option
      shift # past argument
      ;;
    --no-wifi)
      NO_WIFI=yes
      shift # past option
      ;;
    -r|--res|--resolution)
      if [ $# -lt 2 ]; then
        echo "$EXE: usage error: -r or --res missing value" >&2
        exit 2
      fi
      RESOLUTION="$2"
      shift # past option
      shift # past argument
      ;;
    --no-resolution)
      RESOLUTION=
      shift # past option
      ;;
    -b|--boot)
      if [ $# -lt 2 ]; then
        echo "$EXE: usage error: -b or --boot missing value" >&2
        exit 2
      fi
      BOOT_DIR="$2"
      shift # past option
      shift # past argument
      ;;
    -q|--quiet)
      QUIET=yes
      shift # past option
      ;;
    -v|--verbose)
      VERBOSE=yes
      shift # past option
      ;;
    --version)
      SHOW_VERSION=yes
      shift # past option
      ;;
    -h|--help)
      SHOW_HELP=yes
      shift # past option
      ;;
    *)
      # unknown option
      if echo "$1" | grep -q ^- ; then
        echo "$EXE: usage error: unknown option: \"$1\"" >&2
        exit 2
      else
        # Use as the SSID

        if [ -n "$WIFI_SSID" ]; then
          echo "$EXE: usage error: too many arguments" >&2
          exit 2
        fi
        WIFI_SSID="$1"
      fi
      shift # past argument
      ;;
  esac
done

#----------------
# Help and version options

if [ -n "$SHOW_HELP" ]; then
  if [ -n "$DEFAULT_COUNTRY" ]; then
    CTRY=" (default: $DEFAULT_COUNTRY)"
  else
    CTRY=
  fi
  
  cat <<EOF
Usage: $EXE_EXT [options] SSID-case-sensitive
Options:
  -c | --country XX      two letter ISO 3166-1 country code$CTRY
  -k | --psk HEXVALUE    Wi-Fi password and SSID encoded as PSK in hexadecimal
  -p | --password VALUE  Wi-Fi password in plaintext instead of prompting for it
       --no-wifi         do not configure Wi-Fi (default: configure it)

  -r | --resolution RES  set hdmi_group and hdmi_mode (default: $DEFAULT_RESOLUTION)
       --no-resolution   do not configure HDMI (default: configure it)

  -b | --boot DIR        mounted location of the Raspberry Pi boot partition

  -q | --quiet           output nothing unless an error occurs
  -v | --verbose         output extra information when running
       --version         display version information and exit
  -h | --help            display this help and exit

HDMI resolutions are "hdmi_group/hdmi_mode" (with "0" for auto-detect):
  e.g. 1/4=720p, 1/16=1080p, 2/16=1024x768, 2/35=1280x1024, 2/51=1600x1200
  https://www.raspberrypi.org/documentation/configuration/config-txt/video.md
EOF
  exit 0
fi

if [ -n "$SHOW_VERSION" ]; then
  echo "$PROGRAM $VERSION"
  exit 0
fi

#----------------
# Other options

if [ -z "$WIFI_SSID" ]; then
  echo "$EXE: usage error: missing SSID (-h for help)" >&2
  exit 2
fi

if [ -z "$NO_WIFI" ]; then
  # Setup Wi-Fi
  
  if [ -z "$WIFI_COUNTRY" ]; then
    echo "$EXE: usage error: missing Wi-Fi country code" >&2
    exit 2
  fi
  WIFI_COUNTRY=$(echo $WIFI_COUNTRY | tr a-z A-Z) # uppercase
  if ! echo "$WIFI_COUNTRY" | grep -q '^[A-Z][A-Z]$' ; then
    echo "$EXE: usage error: bad two-letter country code: \"$WIFI_COUNTRY\"">&2
    exit 2
  fi

else
  if [ -n "$WIFI_PASSWORD_PLAINTEXT" ] || [ -n "$WIFI_PASSWORD_PSK" ] ; then
    echo "$EXE: usage error: passwords are not needed with --no-wifi">&2
    exit 2
  fi
fi

HDMI_FORCE_HOTPLUG=
HDMI_GROUP=
HDMI_MODE=

if [ "$RESOLUTION" = '0' ]; then
  # Group 0: auto-detect from EDID (no hdmi_mode)
  HDMI_FORCE_HOTPLUG=
  HDMI_GROUP=0
  HDMI_MODE=

elif [ -n "$RESOLUTION" ]; then
  if ! echo "$RESOLUTION" | grep -q '^[1-9][0-9]*/[1-9][0-9]*$' ; then
    echo "$EXE: usage error: resolution is not \"GROUP/MODE\": $RESOLUTION" >&2
    exit 2
  fi

  HDMI_FORCE_HOTPLUG=1
  HDMI_GROUP=$(echo $RESOLUTION | sed 's/\/.*$//')
  HDMI_MODE=$(echo $RESOLUTION | sed 's/^.*\///')
  
  if [ $HDMI_GROUP -eq 1 ]; then
    # Group: Consumer Electronics Association (CEA) i.e. televisions

    if [ $HDMI_MODE_MAX1 -lt $HDMI_MODE ]; then
      echo "$EXE: usage error: hdmi_mode exceeds maximum of $HDMI_MODE_MAX1: $RESOLUTION" >&2
      exit 2
    fi

  elif [ $HDMI_GROUP -eq 2 ]; then
    # Group 2: Display Monitor Timings (DMT)
    
    if [ $HDMI_MODE_MAX2 -lt $HDMI_MODE ]; then
      echo "$EXE: usage error: hdmi_mode exceeds maximum of $HDMI_MODE_MAX2: $RESOLUTION" >&2
      exit 2
    fi
  else
    echo "$EXE: usage error: unknown HDMI group (expect 1 or 2): $RESOLUTION" >&2
    exit 2
  fi
fi

if [ -n "$VERBOSE" ] && [ -n "$QUIET" ]; then
  # Verbose overrides quiet
  QUIET=
fi

#----------------------------------------------------------------
# Check for Raspberry Pi microSD card

_has_raspberry_pi_os_files () {
  DIR=$1

  for F in config.txt start.elf bootcode.bin LICENCE.broadcom cmdline.txt; do
    if [ ! -f "$DIR/$F" ] ; then
      return 1 # failed
    fi
  done

  return 0 # passed
}

if [ -z "$BOOT_DIR" ]; then
  # No boot location provided, try and automatically find it

  # Trying these candidate places:
  # - macOS: /Volumes/boot
  # - Raspberry Pi OS: /media/pi/boot
  # - Others?
  #
  # Note: on a running Raspberry Pi, /boot is the boot partition. But
  # this script should not need to update it, since in a running
  # Raspberry Pi there are more direct ways to configure Wi-Fi and the
  # SSH server. If you really want to use this script on the currently
  # running Raspberry Pi's boot partition, use "--boot /boot" to
  # explicity choose it.
  
  for CANDIDATE in /Volumes/boot /media/pi/boot ; do
    if [ -d "$CANDIDATE" ] ; then
      # Directory exists
      # Check some contents are expected for a Raspberry Pi image

      if _has_raspberry_pi_os_files "$CANDIDATE" ; then
        BOOT_DIR="$CANDIDATE"
        break
      fi
    fi
  done

  if [ -z "$BOOT_DIR" ]; then
    echo "$EXE: error: mounted Raspberry Pi boot partition not found (use --boot)" >&2
    exit 1
  fi

else
  # Boot location provided: it must contain a real Raspberry Pi boot partition
  if ! _has_raspberry_pi_os_files "$BOOT_DIR"; then
    echo "$EXE: error: not a Raspberry Pi image: missing files: $BOOT_DIR">&2
    exit 1
  fi
fi

if [ -n "$VERBOSE" ]; then
  echo "$EXE: Raspberry Pi OS boot partition: $BOOT_DIR"
fi

# Check if first boot has already happened or not

ALREADY_BOOTED=
if ! grep -q 'init_resize.sh' "$BOOT_DIR/cmdline.txt" ; then
  ALREADY_BOOTED=yes
fi

#----------------------------------------------------------------
# Constants: files on boot partition to be configured

WIFI_CONF=$BOOT_DIR/wpa_supplicant.conf
CONFIG_FILE=$BOOT_DIR/config.txt
SSH_FILE=$BOOT_DIR/ssh

#----------------------------------------------------------------
# Check permission

if [ ! -w "$BOOT_DIR" ]; then
  echo "$EXE: error: insufficient privileges to write to $BOOT_DIR" >&2
  exit 1
fi

if [ ! -w "$CONFIG_FILE" ]; then
  echo "$EXE: error: insufficient privileges to write to $CONFIG_FILE" >&2
  exit 1
fi

#----------------------------------------------------------------
# Configure Wi-Fi

if [ -z "$NO_WIFI" ]; then

  #----------------
  # Get password
  
  if [ -n "$WIFI_PASSWORD_PLAINTEXT" ]; then
    # Plaintext password provided
    
    if [ -n "$WIFI_PASSWORD_PSK" ]; then
      echo "$EXE: usage error: do not use both --password and --psk" >&2
      exit 2
    fi
  else
    if [ -n "$WIFI_PASSWORD_PSK" ]; then
      # PSK password provided

      # Check value is hexadecimal
      if ! echo "$WIFI_PASSWORD_PSK" |  grep -q -E '^[0-9A-Fa-f]{66}$'; then
        echo "$EXE: usage error: --psk value must be 64 hexadecimal chars" >&2
        exit 2
      fi
    else
      # No password provided on command line
      
      # Prompt for the plaintext Wi-Fi password
      
      until [ -n "$WIFI_PASSWORD_PLAINTEXT" ];
      do
        # Note: -s works in Bash and Zsh, but not in all shells
        if ! read -s -p "Wi-Fi password for \"$WIFI_SSID\": " \
             WIFI_PASSWORD_PLAINTEXT ; then
          echo
          echo "$EXE: aborted" >&2
          exit 1
        fi

        echo

        if ! (echo "$WIFI_PASSWORD_PLAINTEXT" | grep -q -E ^.{8}); then
          echo "Error: wrong length (must be 8 characters or longer)">&2
          WIFI_PASSWORD_PLAINTEXT=
        fi
      done
    fi
  fi

  #----------------
  # Configure Wi-Fi
  
  # https://www.raspberrypi.org/documentation/configuration/wireless/headless.md
  # https://www.raspberrypi.org/documentation/configuration/wireless/wireless-cli.md
  # https://w1.fi/cgit/hostap/plain/wpa_supplicant/wpa_supplicant.conf

  if [ -n "$WIFI_PASSWORD_PSK" ]; then
    # PSK provided

    # Store the PSK (value NOT in quotes means PSK in hexadecimal)
    PSK_ENTRY="psk=$WIFI_PASSWORD_PSK"
  else
    # Plaintext passphrase provided

    if which wpa_passphrase >/dev/null 2>&1; then
      # wpa_password program available: use it to hash the SSID and passphrase
      # wpa_password is installed on the Raspberry Pi OS

      # Store the PSK (value NOT in quotes means PSK in hexadecimal)
      PSK_ENTRY=$(wpa_passphrase "$WIFI_SSID" "$WIFI_PASSWORD_PLAINTEXT" \
                    | sed 's/^[\t ]*psk=/psk=/' \
                    | grep -E '^psk=[0-9A-Fa-f]{64}')
    else
      # Store passphrase as plaintext (value in double quotes means plaintext)
      PSK_ENTRY="psk=\"$WIFI_PASSWORD_PLAINTEXT\"" # value in quotes

      # TODO: find a way to calculate the PSK without using wpa_passphrase
      # http://jorisvr.nl/wpapsk.html
      # PSK = pbkdf2_hmac_sha1(password_, salt_ssid, iter=4096, key_len=256)
    fi
  fi

  cat > "$WIFI_CONF" <<EOF
ctrl_interface=DIR=/var/run/wpa_supplicant GROUP=netdev
update_config=1
country=$WIFI_COUNTRY

network={
  scan_ssid=1
  ssid="$WIFI_SSID"
  $PSK_ENTRY
}
EOF

  if [ -z "$QUIET" ]; then
    echo "$EXE: created: $WIFI_CONF" >&2
  fi

  if [ -n "$VERBOSE" ]; then
    cat "$WIFI_CONF"
  fi

else
  # No Wi-Fi

  if [ -e "$WIFI_CONF" ]; then
    rm  "$WIFI_CONF"
    if [ -z "$QUIET" ]; then
      echo "$EXE: deleted: $WIFI_CONF" >&2
    fi
  fi

  if [ -n "$ALREADY_BOOTED" ]; then
    # The Raspberry Pi has already been booted, so if the Wi-Fi is
    # already set up, the absence of the "wpa_supplicant.conf" file will not
    # disable it.
    echo "$EXE: warning: already booted: Wi-Fi might have been configured" >&2
  fi
fi

#----------------------------------------------------------------
# HDMI_MODE

_set_config () {
  ITEM=$1
  VALUE=$2

  # Use temporary file instead of "sed -i" since the -i option
  # is not portable between BSD and GNU versions of sed.
  
  TMP="/tmp/${PROGRAM}-$$.tmp"
  
  # Revert statement to original commented form, if it has been uncommented
  sed "s/^[ \t]*$ITEM=.*/#$ITEM=/" $CONFIG_FILE > $TMP
  
  if [ -n "$VALUE" ]; then
    # Change commented statement to desired value
    sed "s/^[ \t]*#$ITEM=.*/$ITEM=$VALUE/" $TMP > $CONFIG_FILE
  else
    cp $TMP $CONFIG_FILE
  fi

  rm $TMP
}

#----------------

if [ -n "$HDMI_GROUP" ]; then
  # Configure HDMI

  _set_config hdmi_force_hotplug "$HDMI_FORCE_HOTPLUG"
  _set_config hdmi_group "$HDMI_GROUP"
  _set_config hdmi_mode "$HDMI_MODE"

  if [ -z "$QUIET" ]; then
    echo "$EXE: configuring HDMI to $HDMI_GROUP/$HDMI_MODE: $CONFIG_FILE" >&2
  fi
fi

#----------------------------------------------------------------
# SSH

touch $SSH_FILE

if [ -z "$QUIET" ]; then
  echo "$EXE: created: $SSH_FILE" >&2
fi

#----------------------------------------------------------------
# Finished

if [ -z "$QUIET" ]; then
  echo "$EXE: done"
fi

#EOF
